libraries/AR_Motors/AP_MotorsUGV.cpp

```diff
@@ -213,8 +220,101 @@ void AP_MotorsUGV::set_throttle(float throttle)
         return;
     }
 
+    
     // check throttle is between -_throttle_max and  +_throttle_max
-    _throttle = constrain_float(throttle, -_throttle_max, _throttle_max);
+    float throttle_chk = constrain_float(throttle, -_throttle_max, _throttle_max);
+    
+    if (_esc_backward_unlock == 0)
+    {
+        _throttle = throttle_chk;
+        return ;
+    }
+
+    /*
+    ** ((throttle_chk == 0) && (_esc_backward_unlock_active)) continue unlocking if user releases throttle but it is zero not forward.
+    ** (throttle_chk < 0)  backward mode anyway
+    */
+    if ((throttle_chk < 0.0f) || ((throttle_chk <= 0.0f) && (_esc_backward_unlock_active) ))
+    {
+        const uint32_t now = AP_HAL::millis();
+        
+        switch (_esc_backward_unlock_state)
+        {
+            case UnlockState::UNLOCK_WAITING:
+                _esc_backward_unlock_simulated_signal=0;
+                _throttle = 0.0f;
+                _last_set_throttle_call = now;
+                _esc_backward_unlock_active = true;
+                _esc_backward_unlock_state = UnlockState::UNLOCK_STEP_DEC_TO_MAX_NEG;
+            break;
+
+            case UnlockState::UNLOCK_STEP_DEC_TO_MAX_NEG:
+                if ((now - _last_set_throttle_call) >= 20)
+                {
+                    _esc_backward_unlock_simulated_signal-=2.0f;
+                    _throttle = _esc_backward_unlock_simulated_signal;
+                }
+                _last_set_throttle_call = now;
+
+                if (_esc_backward_unlock_simulated_signal <= - 98.0f)
+                {
+                    _esc_backward_unlock_simulated_signal = -100.0f;
+                    _esc_backward_unlock_state = UnlockState::UNLOCK_STEP_STAY_MAX;
+                }
+            break;
+            
+            case UnlockState::UNLOCK_STEP_STAY_MAX:
```

Comment:
This can be combined with the above step. 
```
_esc_backward_unlock_simulated_signal -= 2.0;
_throttle = fmaxf(_esc_backward_unlock_simulated_signal, -100.0)
```

Reply:
sorryI didnt get it. 
UnlockState::UNLOCK_STEP_DEC_TO_MAX_NEG is value based  i.e. state ends after output reaches -100 by reducing it -2 each step takes 20ms.

UnlockState::UNLOCK_STEP_STAY_MAX is time based delay for 300 ms

I dont see how can we unify these two steps.